### ISSUE TITLE
Proxy class without parameterless constructor

### ISSUE THREAD

**Issue by Melchy** [castleproject/Core#588]
Hi,
I would like to proxy a class which doesn't have parameterless constructor. This proved to be much more challenging than I expected. 

Initially I thought I can create instance using `FormatterServices.GetUninitializedObject` and then use `CreateClassProxyWithTarget`. This didn't work because`CreateClassProxyWithTarget` requires parameterless constructor.

I found hack somewhere on the stack overflow:

```csharp
private class MyProxyGenerator : ProxyGenerator
{
    public object CreateClassProxyWithoutCallingConstructor(Type type, IInterceptor interceptor)
    {
        var proxyType = CreateClassProxyType(type, Array.Empty<Type>(), ProxyGenerationOptions.Default);
        var instance = FormatterServices.GetUninitializedObject(proxyType);
        SetInterceptor(instance, interceptor);
        return instance;
    }

    private void SetInterceptors(object proxy, Interceptor interceptors)
    {
        var field = proxy.GetType().GetField("__interceptors", BindingFlags.Instance | BindingFlags.NonPublic);
        if (field == null)
        {
            throw new InvalidOperationException("Internal castle windsor property - __interceptors not found.");
        }
        field.SetValue(proxy, new IInterceptor[] {interceptor});
    }
}
```
This works well but needs to access internal castle properties using reflection.

My question is - is it possible to add method which can proxy class without parameterless constructor? 

I also found this very old [issue](https://github.com/castleproject/Core/issues/63) which was closed mainly because .net core didn't support `FormatterServices.GetUninitializedObject`.  This method is available since .net core 2.0

**Comment by stakx** [comment]
Perhaps I'm misunderstanding your goal here, but it seems like you're overcomplicating things:

> `CreateClassProxyWithTarget` requires parameterless constructor

This is incorrect! There are several method overloads where you can pass in `constructorArguments` that will be forwarded to (one of) the proxied class' constructors, for example:

https://github.com/castleproject/Core/blob/d88e94447150996973d947a150ffda5efc0c3c1c/src/Castle.Core/DynamicProxy/ProxyGenerator.cs#L951-L952

(`CreateClassProxy` has similar overloads with a `constructorArguments` parameter.)

What DynamicProxy does not allow is omitting `constructorArguments` when proxying a class that only has parameterized constructors; that is, for such classes, you _must_ pass `constructorArguments` that match one of the proxied class' constructors.

In other words, DynamicProxy does not offer anything similar to what `FormatterServices.GetUninitalizedObject` does. IMHO this is a good thing, since `GetUninitializedObject` bypasses any logic in the class constructor that might establish class invariants. That's not something that one should do casually in well-designed code, so I'd say it's unlikely that we would want to support that scenario in DynamicProxy.

> is it possible to add method which can proxy class without parameterless constructor?

See above; we already have overloads taking `constructorArguments`, and in my opinion it would be best not to add anything akin to `GetUninitializedObject`.

**Comment by Melchy** [comment]
Thank you for quick response.
I didn't notice that I can pass arguments to `CreateClassProxyWithTarget`. But this is still not usable in my case. 

I'm writing a nuget package which allows users to proxy any class with any constructor. I could force them to implement parameterless constructor but that is inconvenient.

> GetUninitializedObject bypasses any logic in the class constructor

This is not a problem because my proxy never calls the proxied method.

The package I'm talking about is [here](https://github.com/Melchy/Ridge). The readme is just draft but the examples should provide enough information about the problem I'm trying to solve.

**Comment by stakx** [comment]
> This is not a problem because my proxy never calls the proxied method.

It may not be a problem in the specific scenario you're looking at, but it isn't necessarily true in the general case.

My thinking is that in general, constructor parameters are there for a reason. Constructor parameters can be seen as the minimal inputs necessary for the class to establish a valid initial state (i.e. to establish its class invariants). If you bypass the constructor, you're no longer guaranteed to get a valid initialized object; you're breaking encapsulation.

So in general, bypassing constructors shouldn't be done. Probably the worst thing we could do in DynamicProxy would be to let it bypass constructors silently, i.e. if you haven't specified constructor arguments, but the class constructors all require parameters, then it would silently use `FormatterServices.GetUninitializedObject` behind the scenes to instantiate the object. Users who (like you) haven't seen the method overloads accepting `constructorArguments` would then by default get unfinished objects.

So _if_ we added support to DynamicProxy for bypassing parameterized constructors, it would probably have to be an API that you have to explicitly opt in to.

That being said, in my current opinion, your scenario seems like an edge case that, by itself, doesn't warrant adding such a new API.

I'm happy to look at your requirements more closely once I have a little more time on my hands, perhaps there are other solutions that don't require `GetUninitializedObject` that we haven't considered yet.

**Comment by jonorossi** [comment]
Yes, it does appear that .NET Core 2.0 added `FormatterServices.GetUninitializedObject` back so that developers can build serialization outside of the runtime since binary serialization has been obsoleted. It appears that Microsoft added `RuntimeHelpers.GetUninitializedObject` for this purpose, but because that one isn't available on .NET Framework they also exposed the `FormatterServices` which just calls the `RuntimeHelpers` one.

This topic has been raised in #21 and #63 in the past.

Krzysztof mentions (https://github.com/castleproject/Core/issues/63#issuecomment-139387865) some of the things that make this a bit more complicated, firstly that DynamicProxy populates it own state via constructors. Another problem is that skipping calling the base constructor creates unverifiable IL and creates situations which can crash the runtime.

Throughout DynamicProxy's history we've followed a general rule that we want to create safe and verifiable code, essentially facilitating legal things you can do yourself in C# you write manually. I'm not going to stand in people's way, but I'm not convinced this is a general purpose feature that is worth maintaining in DynamicProxy, especially when users can maintain this in their own code as a pretty small workaround.
